docs/source/enterprise-guide/change-management.mdx

```diff
@@ -0,0 +1,537 @@
+   gateway after the final publish.
+3. Commit the changes to the respective repositories. The next release will be a
+   no-op because Apollo Studio already has these changes.
+
+[rover]: https://www.apollographql.com/docs/rover/
+
+We might choose to keep a copy of subgraph schemas in a separate code repository
+for review and auditing purposes.
+
+Out-of-band publishing has the added benefit of near-instant rollbacks, should
+we need to revert the change to shared types. However, keep in mind that
+rollbacks may cause API breakage for clients that have updated their operations
+to depend on the changes.
+
+## Changes to query planning
+
+Query planning changes involve changes to federation metadata that affect how
+the gateway fetches data from subgraphs without changing the schema itself. This
+requires making changes to more than one subgraph schema. Because it involves
+changes to at least three services (the gateway plus two subgraphs), safely
+rolling out changes and being able to roll them back if we encounter errors is
+critical.
+
+Apollo strongly recommends that we isolate query planning changes from other
+kinds of schema or implementation changes. All services should be already have
+working resolvers to support the new query plans.
+
+A common use-case is migrating ownership of an entity type from a monolith
+service to a microservice. The microservice takes on the responsibility for
+providing key fields for the entity and must be able to handle the traffic.
+
+```graphql:title=Before
+# monolith — "owns" the User type
+type User @key(fields: "id") {
+  id: ID!
+  fieldFromMonolith: String
+}
+# microservice — extends the User type
+extend type User @key(fields: "id") {
+  id: ID! @external
+  fieldFromMicroservice: String
+}
+```
+
+```graphql:title=After
+# monolith — extends the User type
+extend type User @key(fields: "id") {
+  id: ID! @external
+  fieldFromMonolith: String
+}
+# microservice — "owns" the User type
+type User @key(fields: "id") {
+  id: ID!
+  fieldFromMicroservice: String
+}
+```
+
+### Release steps
+
+The simplest release process works as follows:
+
+1. **In the monolith repository**, commit a changeset containing the schema
+   change and corresponding implementation to the release branch, then trigger
+   the release pipeline.
+2. The release pipeline executes: building artifacts, publishing subgraph SDL,
+   and triggering a rolling deploy of the monolith.
+
+   At this point, the gateway continues to use the old query plans.
+   **Composition is broken**, so Apollo Studio does not publish the new
+   supergraph to the gateway.
+
+3. **In the microservice repository**, commit a changeset containing the schema
+   change and corresponding implementation to the release branch, then trigger
+   the release pipeline.
+4. The release pipeline executes: building artifacts, publishing subgraph SDL,
+   and triggering a rolling deploy of the microservice.
```

Comment:
For entity migrations, this may make sense. But for migrating a field from one service to another, I don't think this ordering works...

Take for instance, this current schema:
```graphql
# Service A — "owns" all fields on `User`
type User @key(fields: "id") {
  id: ID!
  fieldThatDoesNotMigrate: String
  fieldToMigrate: String
}
```

If we want to move `fieldToMigrate` to Service B, like this:

```graphql:title=After
# Service A — "owns" `User` entity + some `User`-specific fields
extend type User @key(fields: "id") {
  id: ID!
  fieldThatDoesNotMigrate: String
}

# Service B — "owns" `fieldToMigrate` on `User`
type User @key(fields: "id") {
  id: ID!
  fieldToMigrate: String
}
```

I would expect something like the following:
* Deploy Service B w/new schema ownership + new resolver -> this breaks composition but makes new resolver available immediate
  * Old resolver in Service A still available and able to execute current requests at this time
* Deploy Service A to remove schema ownership + old resolver -> this fixes composition, updates schema, and retires the old resolver

Given the above, it's that much more important that schema changes are always published first to ensure we don't remove the old resolver before the routing updates in the gateway occur. Otherwise, we can introduce a temporary outage while services roll.

It may also be worth calling out that if you customize the gateway and change poll intervals for updating the supergraph schema, you need to make sure the entire ecosystem is updated before a subgraph service actually deploys. Perhaps I'm getting to into the weeds here though?

Reply:
This is a really good example that illustrates the need to have resolvers are deployed before the supergraph updates in the gateway. It's implicit in my example — I assume you have entity resolvers in place in both services before the schema changes — but your example makes it clearer. 👍🏻